Nodejs/Product/Nodejs/Workspace/PackageJsonTestContainerDiscoverer.cs

```diff
@@ -0,0 +1,147 @@
+
+        [ImportingConstructor]
+        public PackageJsonTestContainerDiscoverer(IVsFolderWorkspaceService workspaceService)
+        {
+            this.workspaceService = workspaceService;
+            this.workspaceService.OnActiveWorkspaceChanged += this.OnActiveWorkspaceChangedAsync;
+
+            if (this.workspaceService.CurrentWorkspace != null)
+            {
+                this.currentWorkspace = this.workspaceService.CurrentWorkspace;
+
+                var fileWatcherService = this.currentWorkspace.GetFileWatcherService();
+                fileWatcherService.OnFileSystemChanged += this.FileSystemChangedAsync;
+
+                var indexService = this.currentWorkspace.GetIndexWorkspaceService();
+                indexService.OnFileScannerCompleted += this.FileScannerCompletedAsync;
+
+                this.currentWorkspace.JTF.RunAsync(async () =>
+                {
+                    // Yield so we don't do this now. Don't want to block the constructor.
+                    await Task.Yield();
+
+                    // See if we have an update
+                    await AttemptUpdateAsync();
+                });
+            }
+        }
+
+        public Uri ExecutorUri { get; } = NodejsConstants.PackageJsonExecutorUri;
+
+        public IEnumerable<ITestContainer> TestContainers
+        {
+            get
+            {
+                lock (this.containerLock)
+                {
+                    return this.containers.ToArray();
+                }
+            }
+        }
+
+        public event EventHandler TestContainersUpdated;
+
+        private async Task AttemptUpdateAsync()
+        {
+            if (this.currentWorkspace != null)
+            {
+                var indexServce = this.currentWorkspace.GetIndexWorkspaceService();
+                var filesDataValues = await indexServce.GetFilesDataValuesAsync<string>(NodejsConstants.TestRootDataValueGuid);
+
+                lock (this.containerLock)
+                {
+                    this.containers.Clear();
+                    foreach (var dataValue in filesDataValues)
+                    {
+                        var rootFilePath = this.currentWorkspace.MakeRooted(dataValue.Key);
+                        var testRoot = dataValue.Value.Where(f => f.Name == "TestRoot").Select(f => f.Value).FirstOrDefault();
+
+                        if (!string.IsNullOrEmpty(testRoot))
+                        {
+                            var testRootPath = this.currentWorkspace.MakeRooted(testRoot);
+                            this.containers.Add(new PackageJsonTestContainer(this, rootFilePath, testRootPath));
+                        }
+                    }
+                }
+            }
+            this.TestContainersUpdated?.Invoke(this, EventArgs.Empty);
+        }
+
+        private async Task OnActiveWorkspaceChangedAsync(object sender, EventArgs e)
+        {
+            this.currentWorkspace = this.workspaceService.CurrentWorkspace;
```

Comment:
You're not unsubscribing from the previous workspace's event handlers. Maybe that causes leaks, idk for certain. You will also want to subscribe to the new workspace's file watcher and index service. Each IWorkspace spins up a new instance of these services.